conclave-enclave/src/main/kotlin/com/r3/conclave/enclave/internal/SocketEnclaveHostInterface.kt

```diff
@@ -0,0 +1,237 @@
+            val outgoingMessage = SocketCallInterfaceMessage(
+                    hostThreadID, messageType, callTypeID, payload?.getAllBytes(avoidCopying = true))
+
+            sendMessageToHost(outgoingMessage)
+        }
+
+        fun sendCallMessage(callType: HostCallType, parameterBuffer: ByteBuffer) {
+            sendMessage(SocketCallInterfaceMessageType.CALL, callType.toByte(), parameterBuffer)
+        }
+
+        fun sendReturnMessage(callType: EnclaveCallType, returnBytes: ByteBuffer?) {
+            sendMessage(SocketCallInterfaceMessageType.RETURN, callType.toByte(), returnBytes)
+        }
+
+        fun sendExceptionMessage(callType: EnclaveCallType, exceptionBuffer: ByteBuffer) {
+            sendMessage(SocketCallInterfaceMessageType.EXCEPTION, callType.toByte(), exceptionBuffer)
+        }
+
+        fun handleCallMessage(callMessage: SocketCallInterfaceMessage) {
+            val messageType = callMessage.messageType
+            require(messageType == SocketCallInterfaceMessageType.CALL)
+
+            val callType = EnclaveCallType.fromByte(callMessage.callTypeID)
+            val parameterBytes = checkNotNull(callMessage.payload) { "Received call message without parameter bytes." }
+
+            val returnBuffer = try {
+                handleIncomingCall(callType, ByteBuffer.wrap(parameterBytes))
+            } catch (t: Throwable) {
+                sendExceptionMessage(callType, ByteBuffer.wrap(ThrowableSerialisation.serialise(t)))
+                return
+            }
+
+            sendReturnMessage(callType, returnBuffer)
+        }
+
+        fun initiateCall(callType: HostCallType, parameterBuffer: ByteBuffer): ByteBuffer? {
+            sendCallMessage(callType, parameterBuffer)
+
+            /** Iterate, handling CALL messages until a message that is not a CALL arrives */
+            var replyMessage: SocketCallInterfaceMessage
+            while (true) {
+                replyMessage = messageQueue.take()
+                when (replyMessage.messageType) {
+                    SocketCallInterfaceMessageType.CALL -> handleCallMessage(replyMessage)
+                    else -> break
+                }
+            }
+
+            val replyMessageType = replyMessage.messageType
+            val replyPayload = replyMessage.payload
+
+            check(HostCallType.fromByte(replyMessage.callTypeID) == callType)
+            check(replyMessageType != SocketCallInterfaceMessageType.CALL)
+
+            if (replyMessageType == SocketCallInterfaceMessageType.EXCEPTION) {
+                checkNotNull(replyPayload) { "Received exception message without parameter bytes." }
+                throw ThrowableSerialisation.deserialise(replyPayload)
+            }
+
+            return replyPayload?.let { ByteBuffer.wrap(replyPayload) }
+        }
+
+        /** Handle the initial call */
+        override fun run() {
```

Comment:
If this is only run once, then it's better to not have the `EnclaveCallContext` implement `Runnable`. Otherwise it gives the impression this run method is run periodically or is the main focus of the class. I would suggest rename this to `initialRun` or something and just invoke in on the executor like `newCallContext::initialRun` or similar.